source/language/directives.rst

```diff
@@ -1,14 +1,95 @@
 Directives
 ==========
 
-OpenQASM supports a directive mechanism that allows other information to
-be included in the program. A directive begins with ``#pragma`` and
-terminates at the end of the line. Directives can provide annotations
-that give additional information to compiler passes and the target
-system or simulator. Ideally the meaning of the program does not change
+OpenQASM supports directive mechanisms that allows other information to
+be included in the program. Directives are either pragmas or annotations.
+Both are used to supply additional information to the compiler passes and the
+target system or simulator. Ideally the meaning of the program does not change
 if some or all of the directives are ignored, so they can be interpreted
 at the discretion of the consuming process.
 
+Pragmas
+-------
+
+Pragma directives start with ``pragma`` and continue to the end of line. The
+text after ``pragma`` is a single string, and parsing is left to the specific
+implementation. The ``#pragma`` syntax is deprecated, but is accepted for
+compatibility. Pragmas should be processed as soon as they are encountered; if a
+pragma is not supported by a compiler pass it should be ignored and preserved
+intact for future passes.  Ideally, pragmas should be considered global and
+immutable; avoiding stateful interactions is recommended to prevent unexpected
+behaviors between source files.
+
+Pragmas are useful for extending OpenQASM functionality that is not described in
+this specification, such as adding noise operations.
+
+\*Note: The following examples are simply possible implementations, this
+specification does not define any pragmas. Please consult your tool's
+documentation for supported pragmas.
+
+.. code-block:: c
+   :force:
+
+   rx(0.15) q;
+   pragma noise bit_flip(0.1)
```

Comment:
Maybe a more accurate statement should be "stateless"; they may have positional relevance in an implementation, but should not (not "must not") implement behavior that needs to be enabled and disabled. Annotations essentially provide enclosures around any scoped code. The goal is to avoid issues such as included files in large codebases changing previously defined settings unexpectedly (accidental `translate_off` that isn't closed, changing `simulation_timescale` between files, etc.).

Reply:
This example should be removed - it is better as an annotation.